### ISSUE TITLE
[Bug] BF16 ELU+1 feature map loses positive values through cancellation

### ISSUE THREAD

**Issue by Lexwzp0** [fla-org/flash-linear-attention#1261]
### Checklist

- [x] I have checked [FAQs](https://github.com/fla-org/flash-linear-attention/blob/main/FAQs.md) and existing issues for similar problems.
- [x] This report is in English.

### Describe the Bug

The `feature_map="elu"` branch in [LinearAttention](https://github.com/fla-org/flash-linear-attention/blob/e52dbc0ea19d3a40d7ab7f9eed855d2b473994d2/fla/layers/linear_attn.py#L97-L101) computes `F.elu(x) + 1`. With BF16 inputs in eager mode, the negative ELU result can round to `-1` before adding `1`, losing a positive feature value that BF16 can represent.

For `x=-8`, this returns `0`, whereas direct BF16 `exp(x)` returns `0.00033569336` (FP64 reference: `0.00033546263`). The error is already about 58% at `x=-6`, before the feature becomes zero.

This matters for normalized linear attention: zero features can leave the denominator at its epsilon while ELU backward still propagates a nonzero derivative. The scalar single-token example below gives output `0` and gradient about `3.36e6`; the FP64 reference gives output approximately `1` and gradient about `2.98e-7`.

The first reproduction calls the unmodified upstream feature map on CPU and needs no training setup, data, or Triton kernels. The additional reproduction below exercises the actual FLA layer on CUDA. The report concerns the optional ELU feature map evaluated in low precision in eager mode; it does not concern the default `elementwise_product` feature map, and compiled execution has not been tested.

### Steps to Reproduce the Bug

Reproduced with an unmodified source snapshot of FLA revision `e52dbc0ea19d3a40d7ab7f9eed855d2b473994d2`. The affected file is byte-identical on `main` at `864a87f6ce5be8828bef81eb22baafd41937cdf2` (rechecked September 19, 2026):

```python
import torch

from fla.layers.linear_attn import LinearAttention

phi = LinearAttention(hidden_size=128, num_heads=1, feature_map='elu').feature_map_q
x = torch.tensor([-6.0, -8.0], dtype=torch.bfloat16)
print('FLA feature:', phi(x).tolist())
print('Direct exp:', x.exp().tolist())
print('FP64 reference:', x.double().exp().tolist())

# Scalar, single-token normalized attention: q=phi(x), k=v=1, epsilon=1e-10.
# This demonstrates the consequence without invoking a Triton kernel.
for name, dtype, feature in [
    ('FLA BF16', torch.bfloat16, phi),
    ('Direct exp BF16', torch.bfloat16, torch.exp),
    ('FP64 reference', torch.float64, torch.exp),
]:
    x = torch.tensor(-8.0, dtype=dtype, requires_grad=True)
    q = feature(x)
    output = q / (q + 1e-10)
    dx, = torch.autograd.grad(output, x)
    print(f'{name}: feature={q.item():.8g}, output={output.item():.8g}, dx={dx.item():.8g}')
```

```text
FLA feature: [0.00390625, 0.0]
Direct exp: [0.002471923828125, 0.000335693359375]
FP64 reference: [0.0024787521766663585, 0.00033546262790251185]
FLA BF16: feature=0, output=0, dx=3358720
Direct exp BF16: feature=0.00033569336, output=1, dx=0
FP64 reference: feature=0.00033546263, output=0.9999997, dx=2.9809562e-07
```

The scalar example uses the `1e-10` epsilon in FLA's normalized linear-attention divide step. It illustrates the feature-map error's consequence; it is not a Triton kernel test. PyTorch's individual BF16 operations are rounding as expected.

### Additional layer-level reproduction

This deliberately constructed single-token case uses the unmodified `LinearAttention` layer and its actual CUDA kernels. Its dimensions and projection weights are chosen solely for a synthetic unit-level reproduction. The projections make pre-feature-map `q=-8`, `k=0`, and `v=1` in every channel. With `phi(q)=exp(-8)` and `phi(k)=1`, the FP64 normalized output is approximately `0.99999997365`.

```python
import torch
from fla.layers.linear_attn import LinearAttention

def stable_elu(x):
    return torch.where(x > 0, x + 1, x.clamp_max(0).exp()).to(x.dtype)

layer = LinearAttention(
    hidden_size=128, num_heads=1, feature_map="elu",
    output_norm="identity", do_feature_map_norm=True,
).to(device="cuda", dtype=torch.bfloat16)
eye = torch.eye(128, device="cuda", dtype=torch.bfloat16)
with torch.no_grad():
    layer.q_proj.weight.copy_(-8 * eye)
    layer.k_proj.weight.zero_()
    layer.v_proj.weight.copy_(eye)
    layer.o_proj.weight.copy_(eye)

original_elu = layer.feature_map_q
for name, feature in [("upstream", original_elu), ("stable", stable_elu)]:
    layer.feature_map_q = layer.feature_map_k = feature
    layer.zero_grad(set_to_none=True)
    x = torch.ones(1, 1, 128, device="cuda", dtype=torch.bfloat16)
    y = layer(x)[0]
    y.float().mean().backward()
    print(name, y.min().item(), y.max().item(),
          layer.q_proj.weight.grad.abs().max().item())
```

```text
upstream 0.0 0.0 296960.0
stable 1.0 1.0 0.0
```

Only the feature-map callable is replaced between the two runs; weights, inputs, normalization and kernels stay the same. This is a numerical edge-case reproduction, not a claim that normally initialized models routinely produce this activation pattern.

### Expected Behavior

For negative inputs where `exp(x)` is representable, the ELU+1 feature map should approximate `exp(x)` without first subtracting and then adding `1`. In the example, the normalized output should remain close to `1` with a small gradient.

### Environment Information

- FLA 0.6.0, revision above
- Python 3.12.7; PyTorch 2.9.0+cu128; Ubuntu 22.04.5
- Triton 3.5.0; Transformers 4.57.6
- The reproduction above runs on CPU; the BF16 feature-map regression also reproduces on an NVIDIA RTX PRO 6000 Blackwell.

### Possible Fix

Evaluate the same feature map directly in its two branches:

```python
def elu(x):
    return torch.where(x > 0, x + 1, x.clamp_max(0).exp()).to(x.dtype)
```

The `clamp_max(0)` prevents overflow in the exponential branch for positive inputs, where that branch is unused. The final `.to(x.dtype)` is necessary to preserve the feature-map output dtype under CUDA autocast: `exp` can be promoted to FP32, otherwise Q/K become FP32 while V remains BF16 and the downstream kernel rejects the mismatch. This evaluates the same mathematical feature map.

Rechecked feature values and derivatives on CPU and CUDA in BF16, FP16 and FP32 against FP64 for inputs `[-10, -8, -6, -1, 0, 1, 100]`. The proposed expression's maximum relative error on these feature/derivative probes was approximately 0.002755 in BF16 and 0.0004144 in FP16.

The actual layer was also checked at sequence lengths 1 and 65, with both explicitly BF16 weights/inputs and FP32 weights under CUDA BF16 autocast, in grad-enabled and no-grad execution. The original returned all-zero outputs in these cases; the candidate returned 1 at length 1 and values in `[0.9921875, 1]` at length 65, with finite gradients. The remaining BF16 rounding is not claimed to be eliminated. These are targeted diagnostics, not a claim that the full official test suite has passed. No production kernels or existing test tolerances were changed. I have not benchmarked the candidate or tested `torch.compile`.

These deliberately constructed cases demonstrate a numerical failure mode. They do not establish how frequently it occurs in ordinary workloads or its impact on end-to-end model quality.

**Comment by vickiiimu** [comment]
Thanks for the super detailed writeup. I’d love to work on this if no one has picked it up yet. This would be my first open-source contribution and it seems like a really good issue to learn from. My plan is to start by reproducing the behavior and looking through the relevant implementation/tests, then put together a fix with appropriate regression coverage.

I noticed the issue was closed and reopened a couple hours ago. Is it still open to a fix, and are you already planning to submit a PR? If not, I’d be happy to take it on!

**Comment by Lexwzp0** [comment]
Thanks for your interest in this issue!

I initially closed it because I wasn’t sure whether the behavior was specific to my setup and wanted to investigate further. After additional checks, I confirmed that it could be reproduced independently of my training setup using the unmodified upstream code, so I reopened it.

If you can reproduce it in your environment as well, I’d be happy for you to take it on and work on a fix. Thanks for offering to help!

**Comment by vickiiimu** [comment]
Thanks for confirming! I was able to reproduce the issue on an NVIDIA L40S and have opened #1265 with a fix and regression tests. I also included the before/after benchmarks in the PR. Appreciate the detailed reproducer--it was really helpful!
